src/result.ts

```diff
@@ -922,18 +1002,75 @@ export class Result<Value, Err> {
+	 * which will be returned as-is (the `Ok` types will be merged).
+	 * The operation will be ignored if the result represents a success.
+	 *
+	 * @param transform callback function to transform the value of the result. The callback can be async as well.
+	 * @returns a new {@linkcode Result} instance with the transformed error, or a new {@linkcode AsyncResult} instance
+	 * if the transform function is async.
+	 *
+	 * > [!NOTE]
+	 * > Any exceptions that might be thrown inside the {@link transform} callback are not caught, so it is your responsibility
+	 * > to handle these exceptions. Please refer to {@linkcode Result.mapCatching} for a version that catches exceptions
+	 * > and encapsulates them in a failed result.
+	 *
+	 * @example
+	 * transforming the value of a result
+	 * ```ts
+	 * declare const result: Result<number, Error>;
+	 *
+	 * const transformed = result.mapError((error) => new CustomError('Custom error message')); // Result<number, CustomError>
+	 * ```
+	 *
+	 *
+	 * @example
+	 * doing an async transformation
+	 * ```ts
+	 * declare const result: Result<number, Error>;
+	 *
+	 * const transformed = result.mapError(async (error) => new CustomError(error.message)); // AsyncResult<number, CustomError>
+	 * ```
+	 */
+	mapError<ReturnType>(transform: (error: Err) => ReturnType) {
+		return (
+			this.failure
+				? Result.runError(() => transform(this._error))
+				: isAsyncFn(transform)
+					? AsyncResult.ok(this._value)
+					: this
+		) as ReturnType extends Promise<infer PromiseValue>
+			? PromiseValue extends Result<never, infer ResultError>
+				? AsyncResult<Value, Err | ResultError>
+				: AsyncResult<Value, Err>
+			: ReturnType extends Result<never, infer ResultError>
+				? Result<Value, Err | ResultError>
+				: Result<Value, Err>;
+	}
+
 	/**
 	 * Like {@linkcode Result.map} it transforms the value of a successful result using the {@link transform} callback.
 	 * In addition, it catches any exceptions that might be thrown inside the {@link transform} callback and encapsulates them
 	 * in a failed result.
 	 *
 	 * @param transform callback function to transform the value of the result. The callback can be async as well.
+	 * @param errorTransform optional callback function to transform the value of the error. The callback can be async as well.
 	 * @returns a new {@linkcode Result} instance with the transformed value, or a new {@linkcode AsyncResult} instance
 	 * if the transform function is async.
 	 */
-	mapCatching<ReturnType>(transform: (value: Value) => ReturnType) {
+	mapCatching<ReturnType, ErrorType extends AnyValue>(
+		transform: (value: Value) => ReturnType,
+		errorTransform?: (error: unknown) => ErrorType,
+	) {
 		return (
-			this.success ? Result.try(() => transform(this._value)) : this
+			this.success
+				? Result.try(
+						() => transform(this._value),
+						(error: any) => (errorTransform ? errorTransform(error) : error),
```

Comment:
If I'm correct, you can pass the reference to `errorTransform` fn as is, because they both have the same signature.